### PR TITLE
Code discussions

### PR DESCRIPTION
**Pull request by nicksnyder**
This is a reboot of https://github.com/sourcegraph/sourcegraph-vscode/pull/15 with some differences:
- It doesn't include the copied graphql files. I am happy to make changes here to improve the error handling, etc., but I don't want to create an implicit dependency in the short term.
- It does include updates to use the new API so we don't have to lookup repos by ID.
- It is hidden behind a feature flag.

I tested by pushing this to a separate extension ID and, it doesn't work (our APIs never get called, unlike in development), but it also doesn't break anything. I asked for guidance on how to get it to work here: https://github.com/Microsoft/vscode/issues/53598#issuecomment-426889693

I think this is mergeable pending any review comments (and I would like to get it merged so when we do figure out what is required to make it work, I can do that quickly).

### CODE REVIEW

File: package.json
Comment (emidoots) [inline package.json:13]: in launch.json we are specifying `--enable-proposed-api` and we're also specifying it here. Which one is actually required for us today?

File: package.json
Comment (emidoots) [inline package.json]: We should make it clear that this is not required for regular use of the extension.

File: src/comments.ts
Comment (emidoots) [inline src/comments.ts:31]: Add a comment indicating whether or not these ranges are  _where you can comment_ or _which ranges have a comment_. It seems like the former, but I'm not sure. Is it?

File: src/comments.ts
Comment (emidoots) [inline src/comments.ts:62]: `repoInfo` takes a file path not a file name, this is wrong I believe.

File: src/comments.ts
Comment (emidoots) [inline src/comments.ts:101]: You need to ignore threads whose `thread.target.__typename !== 'DiscussionThreadTargetRepo'` or else this will break when I add more in the future.

File: src/comments.ts
Comment (emidoots) [inline src/comments.ts:122]: `repoInfo` takes a file path not a file name

File: src/comments.ts
Comment (emidoots) [inline src/comments.ts:166]: `repoInfo` takes a file path not a file name

File: src/comments.ts
Comment (emidoots) [inline src/comments.ts]: I think you should write the targetRepo as an inline object here instead of as a separate `targetRepo` variable.

File: src/comments.ts
Comment (emidoots) [inline src/comments.ts]: You need to specify a revision here otherwise as soon as someone pushes a new commit to this repository your threads will become detached from the Git history and it will suck for the web UX later on (you won't be able to jump to the exact code that was commented on).

File: src/comments.ts
Comment (emidoots) [inline src/comments.ts:130]: Either we need to parse the titles out here now (like we do in the webapp), or I need to fix title handling on the backend before we start using this to create discussions on dogfood/prod.

File: src/comments.ts
Comment (emidoots) [inline src/comments.ts:150]: `relativeRev` here should be the exact Git revision the user is viewing, not the branch name. The branch name would return the selection relative to the server's current HEAD revision of the named branch which is not want you want and will make the discussion threads randomly placed in the wrong locations in files if your local Git is behind.

File: src/comments.ts
Comment (emidoots) [inline src/comments.ts:178]: Same thing here, `relativeRev` shouldn't be the branch name

File: src/comments.ts
Comment (emidoots) [inline src/comments.ts]: Yeah let's just show the username here, not the display name. You have to @mention usernames so it makes sense to display those. The only situation where display names make sense are e.g. on hover over a username (which AFAIK this VS Code API does not provide).

File: src/comments.ts
Comment (emidoots) [inline src/comments.ts:214]: I believe the capturing of context lines here is wrong. This is one of the issues I saw in your PR before but I didn't have time to investigate it.

Additionally, before we start using this against dogfood/prod I need to figure out whether or not you should actually be specifying potentially modified context lines here AND specifying an exact Git revision, or if you should be doing something else in that situation. Otherwise, this will leave our data in an inconsistent state and it will be hard to deal with.

File: src/graphql.ts
Comment (emidoots) [inline src/graphql.ts:44]: It would be good to tack on `utm_source=editor` to these URLs so we can know if requests are coming from editors or something else in the future.

**Review by emidoots — CHANGES_REQUESTED** [review]
I left a lot of comments inline, but didn't test this yet.

I am eager to get this merged, too, but I ask that we make sure to address _all_ of my comments here before merging (or get my final approval). I will not block this PR on the GraphQL stuff.

I'll do my best to unblock on the more difficult points that require me to think through how discussions in general handles things, but seriously merging before we address these would be bad (https://github.com/sourcegraph/sourcegraph-vscode/pull/19#discussion_r222555125 and https://github.com/sourcegraph/sourcegraph-vscode/pull/19#discussion_r222553287)

**Comment by emidoots** [comment]
About these two points from your prior PR:

> - [ ] Does not handle local modifications to files (should limit commenting range to non-modified ranges?)
> - [ ] Add logic to refresh known threads periodically

I think refresh logic is not needed. Let's just figure out the current behavior (do I have to close the file and reopen it? close VS Code entirely? etc) and document that somewhere in the README.

Not handling local modifications to files is also OK when reading comments, but we will need to figure out what to do about submitting comments on files that are modified and not comitted, this is what my comment https://github.com/sourcegraph/sourcegraph-vscode/pull/19#discussion_r222555125 is about.

File: src/comments.ts
Comment (felixfbecker) [inline src/comments.ts]: I would recommend to check for falsiness because this would likely get refactored to return `undefined` at some point instead of empty strings

File: src/comments.ts
Comment (felixfbecker) [inline src/comments.ts]: use `thread.comments.nodes.map()`

File: src/graphql.ts
Comment (felixfbecker) [reply src/graphql.ts:44]: We already know that from the `User-Agent` header, which seems like the better way to do this for the API compared to a query param (which is suitable for links cause it's the only way)

**Review by felixfbecker — APPROVED** [review]

File: src/graphql.ts
Comment (emidoots) [reply src/graphql.ts:44]: @felixfbecker What exactly is the User-Agent for Visual Studio Code and how is it distinguishable from other browsers?

File: src/comments.ts
Comment (nicksnyder) [reply src/comments.ts:214]: > I believe the capturing of context lines here is wrong.

Do you mean the logic is wrong, or do you mean we shouldn't be capturing them at all?

VS Code does have a case that web doesn't have to deal with: unpushed (or even unsaved) modifications. If we only want to support discussing committed/pushed code (which is fine), then there is some work to be done to limit the commenting ranges (as discussed in another comment).

File: src/comments.ts
Comment (nicksnyder) [reply src/comments.ts:31]: added a comment

File: src/comments.ts
Comment (nicksnyder) [reply src/comments.ts:130]: Yeah this was a todo I forgot about. My preference is for "title" to be optional and for the backend to handle creating the title in that case.

**Review by nicksnyder — COMMENTED** [review]
Addressed some of the feedback. Will finish the other later today.

File: src/graphql.ts
Comment (emidoots) [reply src/graphql.ts:44]: Got it, I missed that line, thanks!

File: src/comments.ts
Comment (emidoots) [reply src/comments.ts:130]: Yeah I agree, I'll try to make this change tomorrow or over the weekend.

File: src/comments.ts
Comment (emidoots) [reply src/comments.ts:214]: When I last tested, the logic was wrong in some way (but I forget exactly how. I will check / reconfirm this logic is solid and that we're capturing what I expect here. I'll comment back here once I've done this.

I've been thinking about the issue of unpushed/unsaved modifications, and I see a few options (in no particular order):

1. Do not allow commenting on unpushed/unsaved modifications. I assume this would be hard to implement. It would also be unfortunate / not a good long term solution as it would prevent e.g. asking questions about code you are actively developing (something we could immediately support with e.g. email notifications).
2. Allow commenting on unpushed/unsaved file modifications, but clearly denote it by e.g. passing two new fields to the backend when we create threads indicating "this is on code that is modified in the editor" and "this is code that is committed to Git, but may not be pushed to the remote". This lets us e.g. warn the user in the web UI. This seems like the most ideal outcome, but I am unsure how hard detecting these two cases would be. For the first flag, we would need to ask VS Code if the file is modified in their editor somehow (even just a single "is modified" boolean would do, we don't need to do individual line ranges). For the second flag, we can detect this on the backend when creating a comment with a Git revision that doesn't exist in the repository to our knowledge yet.
3. Don't do anything here yet, and just pass a flag to the backend when creating comments indicating "this comment came from VS Code". This would let us defer the decision until later while not e.g. harming the web experience.

Of the above I think option 2 would be the best, and probably relatively simple to implement.

Comment (nicksnyder) [reply src/comments.ts:214]: > Of the above I think option 2 would be the best, and probably relatively simple to implement.

I agree and I wrote similar logic before in the old Sourcegraph Editor comments implementation. I think it reduces to sending the newest commit hash that the selected lines blame to. If this includes uncommitted lines, then the hash is either omitted or uses a sentinel value `00000000` like what git does. If some lines are committed but not pushed, then the web can detect that (e.g. revision won't be found). This gracefully handles comments on commits that aren't yet pushed but will be.

File: src/comments.ts
Comment (emidoots) [reply src/comments.ts:62]: So you are right! I hate "file name" naming :P

File: src/comments.ts
Comment (emidoots) [reply src/comments.ts:214]: Using the sentinel value `00000000` SGTM, but we should document this in the GraphQL API since we will only accept one literal string value (e.g. not `000000000000000000000000000000000000000000` or a different number of zeros like Git does). I will document this. Are you aware of any online Git documentation about this value that I can link to? I couldn't find any.

> If some lines are committed but not pushed, then the web can detect that (e.g. revision won't be found).

I agree, this is the best approach. It also means zero work here for that case.

TODO:

- [ ] @slimsag to update the GraphQL API docs to state the sentinel value `00000000` meaning.
- [ ] @nicksnyder to add the logic in this PR to send `00000000` when comment is on uncomitted code, and send correct commit (or `00000000` "for now" as a hack) when commenting on committed but unpushed code.

To merge, I am fine with something as simple as "If the file is at all modified, we send commit `00000000`" but as the web UX improves it would suck if we have lots of discussions created this way, so we should fix this relatively soon / before we start using this a ton.

Comment (nicksnyder) [reply src/comments.ts:214]: > Are you aware of any online Git documentation about this value that I can link to?

No, but it is basically the commit hash of only zeros so forty zeros is the "canonical" value (everything else is just an abbreviation like you can do with any commit hash).

Do you prefer this sentinel value over just not sending a value at all (the later might be more intuitive from an API perspective)?

Comment (emidoots) [reply src/comments.ts:214]: Jeez, thank you for questioning me on that. This is exactly why I originally made the field nullable, so yeah, we should just send null in that case (we can/should still send the branch name though):

https://github.com/sourcegraph/sourcegraph/blob/1d7722a5295141024f3d2b0e32be790a04827112/cmd/frontend/graphqlbackend/schema.graphql#L387-L389

Comment (felixfbecker) [reply src/comments.ts:214]: I came here to ask too if we can just use null. Sentinel values are bad for type safety.

**Comment by voxxit** [comment]
Any updates on the progress of this feature?

**Comment by sqs** [comment]
It is likely to factor into the big upcoming code modification feature (https://docs.sourcegraph.com/dev/roadmap#code-modification), but it's all still experimental now. If the code modification stuff describes a problem you have and you want to give input into solution, shoot us an email and we'd love to show you.

**Comment by felixfbecker** [comment]
Closing this since code discussions were removed from the Sourcegraph backend.

**Pull request closed without merging**
